### ISSUE TITLE
Final Project

### ISSUE THREAD

**Issue by aisobran** [LambdaConglomerate/x9115lam#17]
We should have a proposal by Tuesday. During class we talked about adding novel heuristics to PSO. I think it's a great idea. We should also consider what models we want to choose. I suggest everything in moea problems. We could also use it to tune machine learning algorithms. This would be pretty straight forward as we could use sci-kit and a data set from kaggle.

Another alternative would be to code up gale and apply these models.

**Comment by meneal** [comment]
I'm good with that.  I guess my only question when we actually talk to Menzies is whether rolling through all of the models in moea would be enough for him.  If it is I think that's a totally straightforward way to roll to just get it all knocked out.  I definitely like the idea of using PSO, do you think we even need to come up with novel heuristics, or would that be just sort of for sport?

**Comment by ghost** [comment]
I agree with using the moea problems with PSO, and I also agree that we should ask Menzies if we should try implementing some heuristics. We should focus on getting everything to work first and then adding heuristics only after that. In our proposal, we should mention that we are considering different heuristics for PSO and discuss with Menzies.

**Comment by meneal** [comment]
Did he even want us to write anything up?

**Comment by ghost** [comment]
I remember him saying something like that last class. I think he just wants a couple of sentences on a piece of paper.

**Comment by aisobran** [comment]
Yeah we need some form of project proposal by Tuesday.

**Comment by meneal** [comment]
I just started a proposal in the final project folder.  I wasn't sure what we wanted for the number of heuristics so I just put an X in as a placeholder.  I used word since I was having some problems with Latex.  El Capitan seems to have messed me up a bit.  How many Heuristics should we have up there?  Or should we just leave it blank and ask him about scope when we talk to him?

**Comment by aisobran** [comment]
Actually could we do markdown?

**Comment by meneal** [comment]
Totally fine with me.  We can just delete that word file.

On Mon, Oct 19, 2015 at 11:46 AM, Alexander Sobran <notifications@github.com

**Comment by ghost** [comment]
I don't think we need to say exactly how many heuristics we want to implement.

Also, a quick question on the predator heuristic. Why would we have the predator keep the swarm away from the global best? It seems we should keep the swarm away from known bad areas of the space, right? Are we trying to avoid local extrema by doing this?

**Comment by meneal** [comment]
Yeah local extrema avoidance was my point with that.

**Comment by meneal** [comment]
I've printed the markdown that was up there for the final project to bring to class tonight.  Do we want to change anything on it based on what we talked to Menzies about on Tuesday?  I can reprint in the library if we need to change.  Also if we're good with this we might as well send to Rahul and Menzies a link too based on Rahul's email from the other day.

**Comment by aisobran** [comment]
Sounds good to me.

**Comment by ghost** [comment]
I've copied our HW6 code to the project folder, so we can begin working on it. I know both of you are busy with data sciences right now, so I'll see if I can get some sort of PSO up soon.

**Comment by aisobran** [comment]
I just wanted to give you guys a heads up that I will code the moea problems using our framework while I'm flying. I should have them up by end of Friday.

**Comment by ghost** [comment]
Matt, are you planning on coming to class today? Menzies is doing the
rounds with all the groups and I would rather have another group member
with me.

On Thu, Nov 5, 2015 at 10:01 AM, Alexander Sobran notifications@github.com
wrote:

> I just wanted to give you guys a heads up that I will code the moea
> problems using our framework while I'm flying. I should have them up by end
> of Friday.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/LambdaConglomerate/x9115lam/issues/17#issuecomment-154085055
> .

## 

_Joseph Sankar (About Me) http://josephsankar.me_
Master of Computer Science, Class of 2016, North Carolina State University
Bachelor's in Computer Science, Class of 2015, North Carolina State
University

**Comment by ghost** [comment]
I just pushed a really basic start of PSO. All the vectors are there, but
they don't change right now because I'm not calculating the fitness of N.

If anyone is planning on coming to class today, please let me know.
Otherwise, I'll stay home and work on this more.

On Thu, Nov 5, 2015 at 1:07 PM, Joseph Sankar jesankar@ncsu.edu wrote:

> Matt, are you planning on coming to class today? Menzies is doing the
> rounds with all the groups and I would rather have another group member
> with me.
> 
> On Thu, Nov 5, 2015 at 10:01 AM, Alexander Sobran <
> notifications@github.com> wrote:
> 
> > I just wanted to give you guys a heads up that I will code the moea
> > problems using our framework while I'm flying. I should have them up by end
> > of Friday.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/LambdaConglomerate/x9115lam/issues/17#issuecomment-154085055
> > .
> 
> ## 
> 
> _Joseph Sankar (About Me) http://josephsankar.me_
> Master of Computer Science, Class of 2016, North Carolina State University
> Bachelor's in Computer Science, Class of 2015, North Carolina State
> University

## 

_Joseph Sankar (About Me) http://josephsankar.me_
Master of Computer Science, Class of 2016, North Carolina State University
Bachelor's in Computer Science, Class of 2015, North Carolina State
University

**Comment by meneal** [comment]
I'll be in class.

**Comment by aisobran** [comment]
Added all the models from moea. I ran them all through SA with no errors and double checked the formulas once. It would be best if someone could review it because the translation could have mistakes.

**Comment by meneal** [comment]
Papers that I'm reading right now to get a grip on everything for our project on this:
- [Original 1995 PSO Paper, Kennedy and Eberhart](http://www.cs.tufts.edu/comp/150GA/homeworks/hw3/_reading6%201995%20particle%20swarming.pdf)
- [PSO, Methods, Taxonomy, and Applications](https://github.com/timm/sbse14/wiki/etc/pdf/09reviewPSO.pdf) 
- [An Off The Shelf PSO](https://github.com/timm/sbse14/wiki/etc/pdf/Off-The-Shelf_PSO.pdf)
- [Scatter PSO](https://github.com/timm/sbse14/wiki/etc/pdf/07psoScatter.pdf)
- [Pursuit-Escape Particle Swarm Optimization](https://drive.google.com/open?id=0B84yL10MvrUIWkduYTJWR3Y4VDA)
- [Optimized Fuzzy Clustering By Predator Prey Particle Swarm Optimization](https://drive.google.com/open?id=0B84yL10MvrUIdDdSeTE1aGphWjg)

So, The Off the Shelf article saw one of the most useful.  It really presents PSO in an easy to understand form, and it probably will be pretty helpful in terms of setting up all of the different  tunings for this thing.  The last two are really just examples of what we're already trying to do, I haven't read them yet.  I pretty much went through the Taxonomy paper and then just pulled out the papers that seemed most like what we are trying to do.

Neither of you by any means needs to read these papers, or definitely not all of them.  I figured though at least for me it would help me think about what the best architecture would be for this whole setup.  Scatter probably isn't going to be that useful.  I haven't read it or the last two yet.

**Comment by meneal** [comment]
One other thing on this.  We're going to have to do some sort of write up for this anyway, so any lit review we do is going to be useful when it comes to the write up.  If either of you find anything else you think we should read, in Taxonomy or elsewhere, please post it.  I'll read that too.

**Comment by meneal** [comment]
Read those last two papers.  They are terrible.  I'll just put that out there.  Scatter PSO is really interesting though.

**Comment by meneal** [comment]
Here are a few things I'm thinking we'll need to decide to work on this thing.  I'm going to get working on it and will implement stuff as occurs to me/is easiest.  I'm totally happy to change things around as we go though.  I'll post info on pushes that I make on this.  Here are my thoughts before really getting rolling:
- We need some sort of object to keep the velocity, position, and position best for each particle.  I think that object should actually live in the PSO module, but be a separate class.  
- We need a mechanism to operate on the position of the particle.  The thing with that is that it will need to reference back to the constraints and the range of each variable. I'd think we can make the position change a part of the particle, and then have the model check the bounds and constraints.  That's what I'm going to try first.
- We actually need to vary the particle operation mechanism.  This is where the heuristics and major tweaks for the optimizer come in.  My first thought on this is to have some sort of flag that can indicate a  particular type of calculation, where the different calculations live in the particle class.
- In terms of operations there are the following:
  - We can use the standard way which just uses the particle's memory of best and the global best to adjust the movement.
  - We can add the constriction factor K to the first option, or leave it out. 
  - We can add a third factor aside from the global best and particle's memory of best a la the scatter paper from above.
  -  Can split the swarm into 3rds and then put 1/3rd to work as the predator and 2/3rds to work as the prey.  This on it's own has a lot of different flexibility as mentioned in the papers on predator prey.
- The neighborhood is another really debatable aspect of this optimizer.  Looking through the papers listed above, there are a few main ways to do this.  The simplest is just to use the entire swarm as the neighborhood.  I'll start with that, but we could divide the particles into something like pairs or tuples and then have that be the neighborhood, can split the swarm into bigger groups too.  There's a ton of ways to do this.  Using a more complex neighborhood representation would be required to work with the predator prey model.
- Predator prey considerations:  One main consideration is how we actually manipulate the predator and prey for that interaction.  Can have that movement be in a variable number of dimensions.  Apparently  using all of the dimensions for that movement results in horrible performance.  Which isn't a big surprise.
- Menzies said we need to change our calculations to continuous domination from our binary domination representation.  His words "From Hell sucks".  He references the gadgets code for continuous domination.  I haven't looked at it yet, but that's a reasonable thing to look at for this I'm sure.

**Comment by meneal** [comment]
Here's another paper that gives an intuition about initializing velocity:  [Particle Swarm Optimization: Velocity Initialization](http://ieeexplore.ieee.org.prox.lib.ncsu.edu/stamp/stamp.jsp?tp=&arnumber=6256112)

**Comment by meneal** [comment]
Just pushed some initial stuff on creating the cans.  I've pretty much pulled everything else from the pso file.  I don't totally understand the ramifications of using continuous domination yet so I'm going to read that [Zitler](http://www.simonkuenzli.ch/docs/ZK04.pdf) paper so I can get a decent idea of what to do.  

I guess the big thing is that we were using our from hell aggregation as our energy value, so now I guess we use continuous domination to create that energy value, but I just don't know exactly how that works after reading what Menzies has on the site.

**Comment by ghost** [comment]
- So are you planning on not having the velocity, position, and position best as part of s in state.py? These variables only really matter with PSO, although they are a part of each candidate. I can see why you'd want them to be in pso.py, although I'm not sure if they should be their own class.
- I think that's fine, since we can limit based on Xmin[i], Xmax[i], and the constraints. But what should happen if the new position doesn't match the constraints? How do we mutate the candidate without violating the swarming principles?
- For implementing the different particle operations, maybe we can apply them on top of the standard PSO logic. I'm sure we don't need to implement all of these, but it would be nice if our framework would easily allow them to be added in the future.
- The off-the-shelf PSO paper does not recommend having a local neighborhood at all. I think we should start with having just a global neighborhood as you suggested and then go from there if we can.
- I agree that having movement in all decision space would be highly inefficient. You were talking about doing different kinds of transformations yesterday from the decision space into the PSO space. 
- I definitely think we should implement continuous domination for this as Menzies mentioned in class on Thursday. 

Maybe one way to move forward would be to implement the basic PSO from Menzies' notes with the parameters as in the off-the-shelf paper and then add on as we need to. If we do this, then we can compare the results with the basic model with one where we implement predator-prey and local neighborhoods and see if we get better results. How does that sound?

**Comment by ghost** [comment]
Just pushed some really small fixes to pso.py. First I changed the velocity so that each element in init.pos would have its own velocity. 

Going with the idea of initially having just a global neighborhood, I made L = c.pbest, although we might actually want to just make phi_2 = 0 for now.

I also added in the basic updates to velocity and position from Menzies' notes. I made np = 1 so we can just focus on getting the calculations correctly for one particle. You can see that right now the update is not changing anything because the particle is already at its best position. We can change this by giving it a small initial random velocity in each direction or by changing pbest to the results of a base runner. 

Again, this is just the basic PSO implementation, since we want to make sure we have a solid base before we add the fun stuff.

Let me know if you have any questions or concerns.

**Comment by meneal** [comment]
I did decide to put the cans in PSO.  I just made a class for them, but we can always pull that if we decide to go another direction.  It just didn't make sense to put them in the state after I really thought about it.  This way we keep the state more general. 

If the new position doesn't match the constraints I'd think that we would need to do something like flip the velocity's sign on one or multiple variables and then recalculate.  That's the first thing I could think of anyway.  Do you have any ideas off hand?

My thought on this was exactly as you say, to create this with the simplest mechanism possible and then add in different elements to build it out (predator/prey, etc).  The biggest thing for this right now I think is to get continuous domination working right and then be able to build out the rest of the basic framework.  I'll read the cdom article today and hopefully be able to get something working as far as an implementation today or tomorrow.

**Comment by ghost** [comment]
I agree with keeping the state general and putting the PSO variables within PSO if possible. I notice that now we're not using state at all in PSO. Is this intentional, or do you plan on including it? I ask because if we want to include it, we'll need a way to bridge `s` in state.py to the `can` class in pso.py. 

For matching the constraints, I think that would be one way to go about it. Without knowing anything about the constraints, I think doing something stochastic would be a good way to go for now. Maybe we could have a life or patience counter when doing this, because we don't want to keep messing with the candidate. If the candidate runs out of lives, maybe that means it's in a bad decision space? Does it make sense to keep the candidate in that general area anymore? Maybe there's something we can do to move the candidate into an area with better performing candidates where we don't have to retry as often. I'm not sure about it, just throwing some suggestions out there.

I agree with implementing cdom next. Menzies already has some pseudocode up there, but it doesn't look like it'll be that hard to implement. We'll probably just have to adjust `loss` to get the objective values. Let me know if you run into trouble and I'll have a closer look.

**Comment by meneal** [comment]
I do intend to add in the state.  I'm really looking only at prelim stuff right now just to test out some of the machinery of this.  S can just be a vector of cans, at least that was my intention.

As far as a candidate that doesn't meet constraints anymore, I'm really thinking that you just kill that candidate, and regen a new one with a new set of velocity and whatnot.  That's prob the first stupid attempt to roll with.

I have cdom working in simulated annealing now.  I just pulled everything energy related out of the sa code, in another file called sac.  It's working, but its really hard to know if it's working correctly.  The energy figures that we had before really gave us an idea of what was actually going on in some sort of understandable context, without those values we don't have metrics that we can use as easily.  

For example: I ran both sa and sac against the osczyka and fonseca models and just pegged the energy value at 1.0 for sac, since with cdom it's completely a meaningless figure.  I wanted to see if the final candidates were close to each other.  But in reality they probably won't be that close to each other since it's a total frontier.  I'm thinking at this point that we may well need to implement a hypervolume and spread metric so we can get a sense of what's actually happening and whether we are actually coming up with reasonable values.  I suppose we could calculate energy by hand for what we have and check it that way, but we'll need some goodness metric later anyway.  The values I got out of the two different types of optimizers are in a file called sacvsa.txt. 

I've figured out how to include cdom in pso, I just want to make sure that it works before we get any further on what we're doing.  Either of you have any idea how you would calculate spread or hypervolume?

Ignore test2.py.  I just pushed it by accident.  I needed some way to test the model changes I made.

**Comment by meneal** [comment]
Came up with a better idea than trying to figure out how to calculate hypervolume and spread right off the bat.  Just added back in the mechanisms for calculating energy to sac.py, and used only cdom to actually optimize.  I'm coming out with reasonably similar energy values for a final now.  They're a bit different:
- CDom Osyczka2 EBO: 0.151 
- FromHell Osyczka2 EBO: 0.119
- CDom Fonseca EBO: 0.703
- FromHell Fonseca EBO: 0.624

I would say that it's fairly likely that cdom just needs to run longer than from hell to come up with a smaller energy value.  Also we're testing it on a metric that It really doesn't optimize for anymore.  Still it at least gives me a bit of confidence that we have something that works.

I still think it's important to come up with a way to calculate hypervolume and spread though.

**Comment by meneal** [comment]
Made a bunch of changes to the code for PSO.  There were a few mistakes.  Random.uniform takes it's bounds as it's parameters.  For example `random.uniform(0,1)` produces a random value in the range 0,1.  I think it may have been typo, but we had random.unform having parameters of the values to compute which wouldn't have made sense.  I killed the wider for loop and made both computations into list comprehensions.  So now we basically loop through the candidates, and compute the a list comprehension for velocity and then for position.

I decided that cdom should probably live in the model.  Main reason for that is that it actually calls a method of the model, and anywhere else we put it we're going to have to pass it the model, which seemed like a pain in the ass.

I haven't set up the constraints in PSO yet, mostly because I've forgotten exactly how to check them with the model.  I figured when Sasha returns he can put that in.  

As of my most recent push we have a loop that updates all of the cans, and a loop that runs cdom against all of the cans, updating global best and the personal best for each can.  I haven't set things up to work with the state object yet, but I will after it's running with the constraints.  I haven't made any moves on the fitness calculation yet, but I think we may have a meeting with Menzies in our future figure out a good metric to use.

**Comment by meneal** [comment]
So there were a bunch of bugs in this.  But I think a good number were fixed this morning:
- I wasn't checking bounds.  Fixed that.
- There was some sort of aliasing issue with the best candidate selection in cdom.  Fixed that.

The first was obvious, but the second was really hard to fix.  Anyway, here's an idea of where we're at now:

Since I've fixed cdom, heres how its working:  We select the first can in the list and then run it across all of the candidates in the list and output the winner of that run.  We then continue doing that for all of the cans in the list, basically to see if we end up with different cans that dominate at the end of each run through.

So what's interesting is that you can see that the best id is different for many of the rounds in this output:

```
=======================
BEGIN DOM PROC K:  192.0
=======================
ZERO DIFFERENCE IDS: 0 15, diff: 0.008
best id after run  0
ZERO DIFFERENCE IDS: 1 27, diff: 0.043
best id after run  26
best id after run  29
ZERO DIFFERENCE IDS: 3 18, diff: 0.074
ZERO DIFFERENCE IDS: 3 21, diff: 0.021
best id after run  29
best id after run  29
best id after run  28
ZERO DIFFERENCE IDS: 6 23, diff: 0.017
best id after run  28
ZERO DIFFERENCE IDS: 7 17, diff: 0.087
best id after run  28
best id after run  29
best id after run  29
best id after run  29
best id after run  26
best id after run  26
best id after run  29
best id after run  29
ZERO DIFFERENCE IDS: 15 0, diff: 0.008
best id after run  0
ZERO DIFFERENCE IDS: 16 26, diff: 0.074
best id after run  26
ZERO DIFFERENCE IDS: 17 7, diff: 0.087
best id after run  28
ZERO DIFFERENCE IDS: 18 3, diff: 0.074
ZERO DIFFERENCE IDS: 18 21, diff: 0.064
best id after run  29
best id after run  26
best id after run  29
ZERO DIFFERENCE IDS: 21 3, diff: 0.021
ZERO DIFFERENCE IDS: 21 18, diff: 0.064
best id after run  29
best id after run  29
ZERO DIFFERENCE IDS: 23 6, diff: 0.017
best id after run  28
best id after run  29
best id after run  29
ZERO DIFFERENCE IDS: 26 16, diff: 0.074
best id after run  15
ZERO DIFFERENCE IDS: 27 1, diff: 0.043
best id after run  26
best id after run  27
best id after run  28
final best id  28
```

I wanted to control for the fact that they may well have one of the two following features:
- They're in the same position.
- They have almost the same loss per the cdom function.

In the output I have printout for when particles are in the same position (But not some epsilon of the same position, which may be the case), and I have printout when they are within an epsilon of the same loss value.  Notice ZERO DIFFERENCE.  That line first indicates the two IDS that it has decided have a zero difference in loss and then indicates what that difference actually is.  The epsilon level is set to 0.1 for now.

I'm not totally sure what's happening yet.  Just thought I would post some results to what I'm looking at right now.

Also note that at the end of the output I'm indicating the number dead/percent attritiion.  That value indicates the number of particles that die during the run.  Death occurs when they either are outside of bounds, or outside of constraints.  There are a number of other ways to handle out of bounds and out of constraints.  This is just a first shot at it.  If you want to scan through all of the output, it's in console.txt for now.  I'm not using the state logger yet.  I know we'll need to make some changes to make that work and I haven't had time to look at doing that yet.

One other thing: after talking with Sasha a bit yesterday, I think that we should just use energy for a metric for now and then integrate more interesting metrics after talking with Dr. Menzies in class on Tuesday.  I'm concerned about scope, but I think we can get more done before we speak to him.  Also I'm sorry about the number of print statements that are either in the code or commented out right now.  I'm just actively debugging.

**Comment by ghost** [comment]
I've added energy as a metric and the energy of the best solution is added at the end. The energy values are in the right part of the energy spectrum but they look a bit high to me. 

I also made a slight change on line 164 where we decide whether we are dominated by another candidate. If I understood Menzies correctly then just checking that we do not dominate another candidate does not imply that the other candidate dominates us. So I added another check, so we now only update the best when can dominates c and c does not dominate can. Let me know if I'm misunderstanding this part.

I think we have a good framework up now, and we have to start tweaking different things to get lower energy values. We'll probably also need to have more detailed output like SA. It's hard to represent this data in text form because this is really positional data that should be drawn somehow. Anyone have any ideas?

**Comment by meneal** [comment]
So for your first comment:  Energy for our setup is determined dynamically by the seen values for each decision over the entire runtime of the rig.  The code checked in only initializes the maxs and mins, but doesn't actually update them during run time.  This is really part of the reason I wasn't sure I wanted to implement energy for our setup.  Basically we need to either update energy at each step for all of the cans, or we could potentially update max/min for some subset of the cans.  I hadn't completely decided on which way to roll, but updating max and min for 30 cans every iteration could get really expensive.  I'll set that up this afternoon when I can get home.

Second comment:  In fact this second check is embedded in the code for cdom.  The diff amount I'm outputting is actually the difference between the two checks.  Take a look in the code for model and check to see if you're satisfied.

Representation is a whole other thing.  I'm not totally sure yet what to do with that.

**Comment by ghost** [comment]
I do update the max and mins on line 137, but I see what you're saying about it getting expensive. I'm not sure how else to do it. All we are doing is a series of comparisons, and array accesses/sets. And a size of 30 is not very large. If our model needs to scale, then I think we would need to optimize a bit more. But for now I wouldn't worry too much about it as long as it runs in a reasonable amount of time on our machines.

I had a look at cdom and you're right. The calls to loss take care of both sides. However, I would stick with the first part of the if-statement instead of the negated part since it is easier to understand.

**Comment by aisobran** [comment]
You have to run update at every iteration with updateObjectiveMaxMin or the energy calculation is pretty much invalid. It's not too expensive, O(number of objectives \* steps \* candidates) which has strictly lower complexity than the optimizers which is O(number of objectives \* steps \* candidates^2). Since we are not even increasing the complexity by constant by including it, expensiveness is not a good reason to avoid it. 

Even if we don't use energy in the optimizer we should still  updateObjectiveMaxMin because we want valid energies for the final results when we are comparing optimizers.

**Comment by meneal** [comment]
Joseph: You're right about updateObjectiveMaxMin.  You do have it in the right place!  Sorry about that.  I think we have we what we need there.  

Sasha:  You're right.

I fixed another weird bug with best.  There was a problem where we were getting different bests in the same iteration.  That's done.  There is still something happening where the particles are very similar in terms of loss.  But I'm not satisfied that I understand when or why that's happening.  I'll mess with it some more though to try to figure it out.  There certainly are a large number of deaths for Osyczka2.  Almost all of the particles die, that's not that surprising though I guess.

For the main PSO I've turned off a lot of the output, just to see what we get in terms of energy.  Looking at .22 for Osyczyka and .69 for Fonseca.  We can switch those out for whatever.  I just didn't have time to try more than that for the moment.  That's just 500 changes and no retries.

You can see in this extra little file comparison_ids.txt where you can see the correct thing is happening in each iteration in terms of domination.

**Comment by aisobran** [comment]
Yeah I'm thinking dying when outside of bounds/constraints is not going to be a good heuristic.

I added a simple graphing capability so we can see where the particles are going. This may help us understand the movements we're implementing. I'll add color for each particle, just need to flesh this out a little more. 

Also, this grapher will only work with models with two decisions because it's only two dimensional. It will work for more but it will just ignore the any decision past the first two.

I ran tanaka and the behaviour looks very clumped. 

I'll implement a graph of the objectives as well.

**Comment by aisobran** [comment]
Also you'll need matplotlib

**Comment by aisobran** [comment]
How far are these particles supposed to move because I'm seeing some huge jumps?

**Comment by meneal** [comment]
I totally agree, it's not a long term heuristic.  Menzies suggests what I have in a comment in the code.  Reset the position of the particle to max or min depending on it's current position, set the velocity to zero, then the obvious thing is that it will be pulled back in to the search area by the global optima.  I also have inline that we can try either wiping the memory of the particle of its best or leaving its best.  I just wanted to get to the point that we're at now.  There were a bunch of bugs.

I haven't looked at the output yet, but a graphical look at this will be super helpful.  As far as large jumps, you have to keep in mind that velocity is not capped.  The constriction factor should restrict velocity, but from what you're saying it isn't doing a great job of that. :)  Menzies suggests clipping to xmax (the bounds for that particular variable) for each velocity, but then kind of dithers from what I read and talks about the constriction factor supposedly helping.  For now there's no vmax, can totally add one though if you want.

**Comment by aisobran** [comment]
Added unique colors for each can.uniq.

**Comment by aisobran** [comment]
Yeah that makes more sense. It may not be that the jumps are huge it's just that there are so many deaths that it looks very random. Maybe we I could track death and then give it a unique color.

**Comment by meneal** [comment]
Dude, mad props on these graphs.  They're really interesting.  Do you have a link to the documentation you used to set it up?

**Comment by meneal** [comment]
btw:  Interesting results on these runs actually.  We're getting pretty low global energy values on some of these.  Tanaka in particular is really low.  There's such a wide variance in terms of the attrition percentages too. some are super low some kill almost all of the particles.

**Comment by aisobran** [comment]
I'm just using the matplotlib docs. 

I added resizing of the dots based on the energy of the vector. Right now, the bigger the dot the better the vector. After seeing the output it looks pretty good. All the clumping of the particles is at the best  energies. I'm going to reverse this scaling(the smaller the dot the be better the energy) and see the resulting output. This may give a better presentation.

**Comment by aisobran** [comment]
I inverted the relationship so smaller energies (better candidates) are smaller dots. I think this looks better.

I also added a graph for energies. Interesting graphs indeed.

After graphing the energies, I've notices a bug in the code. The objectives are not normalizing between 0 and 1. I'm trying to troubleshoot this right now. Something is not adding up as none of the energy calculations are exceeding these bounds. My initial guess are either the model initially passed in to grapher is copied instead of referenced later so all updates to maxmin are not incorporated (I don't think this is how python works), the max min implementation is off (but if this was the case we'd get some really high energy values),  either some vectors are not being checked in to max min tracking, there's some roundoff error occurring, or the normalization calculation is incorrect.

For sanity check, normalization should be (x - min)/(max - min)?

**Comment by meneal** [comment]
For the copying vs referenced later part you can actually use the id function and print out the response inside pso and inside grapher.  The id function returns a unique identifier for each object.  Could help in debugging.  I just went ahead and did that.  They're both the same unique id:

```
id in pso  4333997392
id in grapher:  4333997392
```

I don't want to mess with your mojo, so I won't make any changes, but It could be that we cull the particles before we update the objective max/min on line 137.  It could be that if we update objective MaxMin on line 117 we may come closer.  

That said, the sheer number of hits outside of the range for Tanaka and Srinivas make me wonder if that could possibly be the cause.  The tanaka ones are super messed up for some reason, we have a large number of energy values below zero and at least a good number over one.  Also have a good number that are running negative and above one on Srinivas.  

Per [here](http://stats.stackexchange.com/questions/70801/how-to-normalize-data-to-0-1-range) it looks like the normalization formula is correct.

**Comment by aisobran** [comment]
Just to clarify you mean we have "a large number of objective values below zero"

If the energies are below zero than we have a bigger problem as they could never be below zero the equation is sqrt(x^2 + y^2 + z^2 ...)

**Comment by meneal** [comment]
Yep, objective values, my bad.

**Comment by ghost** [comment]
The reason I updated the objective max/min on 137 is that we might have had to generate new positions if the candidate is either out of bounds or out of constraints. I don't think the objective values are valid if their of these conditions aren't met. Matt, why do you suggest moving the update to line 117? I'm just curious. 

And Sasha, really good work on the graphs! They make it a lot easier to visualize the candidates and the decision space, and would be really helpful when writing our report.

I do have a question though. In tanaka_energiest.png, it seems there are quite a bit of candidates with negative energy. These are actually the objective values before normalization, correct? I just want to clarify because "energies" is mentioned.

**Comment by aisobran** [comment]
The reasoning for the location of updating objective max min is correct. 

The energies cannot be negative. What you are seeing are negative objective values which shouldn't be negative as they are normalized.

**Comment by aisobran** [comment]
Just pushed a troubleshoot for this. All the graphing is commented out for right now. 

The objective max, mins are not updating correctly. You can see the output at the bottom of console.txt.

Still diving deeper into this.

**Comment by meneal** [comment]
I think I may have fixed this.  I'm not going to push it since I don't want to mess up what you've done over the past hour.  I just commented out a bunch of print lines just to get rid of a bunch of debug shit I had in there.  Commented out: 143-145, 179, 182-183.  Then I added an additional update on line 117:
`model.updateObjectiveMaxMin(can.pos)`

I'm getting no output of lesser or greater for objectives at that point.

I think that what's happening is that if you don't update on line 117 you lose whatever objective values you may have gotten from the movement in line 116.  Obviously some of that movement is going to be out of bounds since the decisions will take on values that are out of both bounds and constraints.  The bad thing is that I would think that would artificially inflate values that we'll get for energy using this rig.  That said running this way I got a really similar value for energy from running with that check or without that check.  Also it works.

**Comment by meneal** [comment]
A quick test with Fonseca and Tanaka has values less than min for a few objectives  But they're all within about 0.05 of the minimum value.  Tanaka has severely reduced performance though, or at least it appears so since we're adding in those values that are out of bounds.  I'm getting 0.64 as an global best for it.

**Comment by meneal** [comment]
One thing I just thought of though is that it's pointless to do separate checks for check constraints and check bounds.  It may be possible to create two new vectors by having separate checks, not very probable though.  Just deleting the `if not model.checkConstraints(can.pos):` line changing the first if to `if not model.checkBounds(can.pos) or not model.checkConstraints(can.pos):` Is at least more succinct if not potentially better.

Just figured it out.  You're adding invalid vectors to the list that you're graphing!  Instead of adding the vector on line 116 you should be adding after the check.  It's no wonder why we get working energy when the update is made before bounds are checked and then get fucked up energy when the update is made after bounds and constraints are checked.  Let me know if I can push the fix.

There are still little fuck ups in Fonseca, but they're all less than 0.2 of fonseca being less than the min.

**Comment by aisobran** [comment]
I came to the same conclusion. I'll push the fix up with updated energy graphs, I'll also rename the graph to objectives because I think ti's causing confusion.

**Comment by aisobran** [comment]
Ok pushed with new graphs. We can clearly see the pareto frontiers now. I'm going to take a break for a little bit and get back to this in the afternoon.

**Comment by meneal** [comment]
Agreed that the change to objectives is better.  I know that they are objectives, but every time I open the file I think energy. It's really interesting to see them hit that wall that is the frontier.  Seems like we pretty much have a working implementation at this point.

**Comment by aisobran** [comment]
So I looked this over and compared it to the lit. Everything looks good except I don't see an intertia constant. But that makes sense, explained below.

It looks like you are implementing the adaptive PSO by Clerc and Kennedy(2002) with constriction factor(here called k, chi in the lit) and not the classical PSO by Kennedy and Eberhart which contains an inertia constant and no constriction factor. Correctly, the constriction factor is implemented with no intertia constant. But I think our baseline should be classical PSO, with adaptive PSO (already implemented) as another comparator.  

The main difference being that there is no constriction factor in the update  and the previous velocity should be scaled by some non-negative inertia factor. 

I can implement this if you guys agree.

**Comment by meneal** [comment]
Yep, that's the main difference.  I only implemented this way because of
the suggestions from the off the shelf pso paper.  It suggests that this is
better than the older inertia methodology.   I'm totally fine with using
inertia as a baseline if you'd like.  Constriction is referred to as K in
the literature as well as chi, K is what they call it in the off the shelf
paper.  I read the initial Kennedy paper, but have totally forgotten what
they used for an inertia calculation.

One thing I would say though is that we could make a case for just using
the off the shelf implementation as a baseline and then just going from
there.  They've already made the case that classical PSO is inferior to
adaptive, so I'm not sure that we need to bother doing that work again.  I
guess if you think it won't take you much time and you really feel like
it's worth it go right ahead, I'm just not sure that it adds that much.  To
some degree going with the older weaker version will inflate our relative
success!

We would likely need to implement some hard bounds for XMax/Xmin and
VMax/Vmin to make it work.  Theoretically constriction is helping to keep
our velocity values lower than they would be otherwise, but I don't think
that the inertia value will do anything to help that.

**Comment by aisobran** [comment]
You're right that the constriction factor removes the need for Vmax and the inertia factor and helps reach convergence and is widely agreed as a good method to avoid optimizing Vmax and inertia. But there is also research that shows that this quick convergence inherent to global adaptivePSO may not yield diverse results. (http://users.cecs.anu.edu.au/~ejmontgomery/publications/2011-12_convctrl_pso.pdf) I think we are actually seeing this if you look at all the clumping occurring the graphs.

I like keeping adaptive PSO as a comparator and as a base PSO to add our heuristics. But I'm not so sure we should take the result of "adaptive PSO is the best" from the off-the-shelf paper as gospel. They test on 5 models, each with single variant for the objective, and the no free lunch theorem would lead me to believe that there will be models where adaptive PSO will perform worse (it doesn't seem like we share any models with OTS-PSO). Additionally, implementing classical PSO will give use two more variables to tinker with when applying heuristics. Other PSO implementations have had success with this including adaptive fuzzy PSO(Eberhart and Shi).

Also, we should consider how the constriction factor will affect the implementation of our heuristics. Do we want the prey's convergence to be function of the constriction factor when being chased away or do we want the ability to be chased away at greater speeds if the predator gets close?  Also, for the second heuristic we want greater repulsion as each particle gets bigger. This will be hard to achieve solely by C1 and C2 as they will be scaled back down by chi.  I guess another option is to add a third term the equation that is outside chi's scaling but it's difficult to foresee if chi's utility is degraded by this.

We should also consider how the neighborhood topology will affect convergence. Kennedy has shown that a local neighborhood converges more slowly(requiring more steps) but has higher probability of finding the desired solution than using global neighborhoods for complex problems (https://svn-d1.mpi-inf.mpg.de/AG1/MultiCoreLab/papers/KennedyMendes02%20-%20Population%20Structure%20and%20PSO.pdf). This actually makes sense as complex problems will most likely have many local optima that a global neighborhood implementation will converge to. Once we implement one of Menzies' nasties I think this will pop up.

For right now should probably split pso.py to:
adaptiveGlobalPSO.py
classicalGlobalPSO.py

I can alter adaptivePSO into classical. It should be easier to split these up than to have an option in pso.py, at least at first.

From here we should implement our heuristics giving us 6 different PSOs. 

Then time-permitting we should try implementing a local neighborhood variation.

**Comment by ghost** [comment]
It would be nice to experiment with these different heuristics, but I would recommend talking to Menzies tomorrow and seeing what he thinks and if this is in scope for a masters level project. Especially since we will not only have to implement these heuristics but test them on some nasty models.

**Comment by aisobran** [comment]
Classical global PSO was added you can see the comparison of the results in the out.txt. I'm putting the graphs together now.

**Comment by aisobran** [comment]
Pics up. Looks like the less convergent behavior is pretty evident vs adaptive.

**Comment by meneal** [comment]
I would've responded earlier today but I've been poking around at some other stuff.  I completely agree with you about the fact that we shouldn't consider OTS PSO to be gospel.  For whatever reason that didn't really pop off to me initially.  It's also awesome of you to post those papers that give some background about why classical may be of use.  I was planning on reading the papers earlier today, but just didn't get a chance.  

I think that the plan you've put forward makes sense.  2 heuristics and one adaptive and one classical for each.  I'm not too worried about scope with that.  We'll see about local neighborhood.  I doubt we'll get to it just because I think it could take a long time to implement depending on how we go about it.

As far as the two heuristic applications and whether they'll be affected by choosing classical vs adaptive, I might take a look back at the two papers I posted towards the beginning of this thread.  I've forgotten whether they used the constriction factor or used inertia.  I'll look at that tomorrow and post.  Might at least give us a bit of a signpost on where to start implementing heuristics with this setup.

Looking at the images.  It's interesting that there is a huge difference between the objectives outputs for some models (Sriinvas, Viennet3,Viennet2) , but then some models have barely any differences (Viennet4, Constr_ex).    For a Viennet 2 and 3 it is possible to see in the graphs a much more delineated frontier with features that would be missed totally by adaptive.  Good work on that shit man.

**Comment by aisobran** [comment]
Just pushed an initial implementation of heuristic no. 2 in classical global PSO. Output looks much different than before. Still trying to tinker with it and there are some issues with deleting candidates as they get combined. You can see some of the pics. 

I noticed a bug in the code. I'm not sure if this was in there initially or I accidentally uncommented the line but at line 191 of adaptive, it looks like we are generating a whole new set of candidates at the end of every loop. I didn't see this an any implementation of PSO. I didn't remove it yet just wanted to make sure it's not supposed to be there.

**Comment by meneal** [comment]
Line 191 is to regenerate a clean slate at the beginning of a new retry.  It's hard to see though because of the indentation.  You can see though above it on line 184 that st.k is decremented and then we indent out, for the outer loop.  We haven't really been using retries so it hasn't mattered, but I just put it there for once we start running with retries.

As far as the setup for 185 through 223, is there any particular reason you didn't include this stuff within the block above it?  Is your intention to move the particles first and then do the repulsion?  The only reason I ask, is that at that point we essentially move the particles twice in each timestep.  At least that ones that are within range of another particle.  I guess that could be ok, I'm just not totally sure.  

Perhaps it could be better to aggregate all of the forces on each particle first and then change the position of the particle?

**Comment by meneal** [comment]
Nice thing about aggregating too is that we could potentially weight each aspect that adds in to the velocity of the particle and tweak that to get different results.

**Comment by aisobran** [comment]
That makes sense about the retries.

We definitely should compute everything together. But I wanted to separate it out right now to be able to trouble shoot. There's a lot of assumptions in the calculation and right off the bat I can tell it's probably undershooting. In a couple of iterations it will be combined.

**Comment by ghost** [comment]
More models: http://jmetal.sourceforge.net/problems.html
DTLZ: http://www.tik.ee.ethz.ch/~sop/publicationListFiles/dtlz2001a.pdf

**Comment by meneal** [comment]
[here](https://github.com/ai-se/Spread-HyperVolume) is the repo for the metrics.  From what I can see we will need to create two files from a run and place them in specific directories, then run the scripts to pull out hypervoulme/spread metrics if we use the current setup for the scripts.

We could rip the scripts apart and add them into our current setup, but it's really complicated enough that I'm not sure that it would actually be worth it to rip them apart.  I'll try to add them as a subdirectory to project and then setup a shell script to automate running our rig and then running the metrics.  I'll poke around at that this morning for awhile.  I'll work in the original adaptiveGlobal to start, and setup a frontier in there.  Some of this may well necessitate moving the can class into state, not the worst thing for sure, but not exactly what I wanted to do either.

One other thing is the double loop we talked about a bit yesterday for cdom.  For these small models we aren't getting any zero diff loss amounts, but that could well change once we add in these fatter models.  There's a block in model.py for at least creating output if two cans have similar loss within some epsilon.  That block is between 113 and 121.  It doesn't do anything any differently if they're within the epsilon it just spits out some output.  Whenever we upgrade to the fatter models we should remember to turn that on just to check and see if something is happening there.

I'll post more when I actually get some shit done.

**Comment by meneal** [comment]
I just realized that calculation of objectives as I'm using them in our case relies on having the updates for the objective max min.  I'm using the function on line 109 of model.py, which uses the normalized objective output.  For that to work we obviously need to continuously update the objectivemaxmins or else that value won't be valid.  So we'll need to make sure to keep that stuff in regardless of whether we actually jettison energy or not.  

All of the metrics algorithms as they're set up take only one frontier set and calculate hypervolume and spread, so for now what I'll try to do is take a one retry frontier and get the metrics working, then after that I'll use the median of multiple runs as the value to calculate the metrics against.

**Comment by ghost** [comment]
I noticed that the pareto frontiers in the metrics were in the same format as the reference pareto fronts on the jMetal site, so it looks like they're in a standard format. I don't think it'll be difficult to create these output files for our candidates.

Also, I'm going to use the wiki to keep links to our project papers like you suggested with some notes as to what each paper is about. We could also use the wiki to store links to our readings since they'll be useful to have later on.

**Comment by meneal** [comment]
Median was a dumb idea.  Going to run cdom against each individual particle's best for each run and then take the dominator out of each run for each particle.  Then we sort of have a best of the best frontier.

**Comment by ghost** [comment]
Just pushed an implementation of DTLZ1. DTLZ has a customizable number of objectives and I didn't want to fit each objective into a lambda so I tried to generate them dynamically. I probably made some mistakes, so have a look over it and feel free to condense the objectives if it's possible and still easy to understand.

I did an initial run of DTLZ1 and it seems we're getting really low energy values. I'm a bit confused though since the paper says that cans with all zeros are optimal but we are not getting such values in our global best nor in our frontiers. I'm just assuming something is wrong with the implementation. Let me know what you guys think.

**Comment by meneal** [comment]
I haven't had a chance to look at dtlz yet.  But I did want to post on what I've done so far with hypervolume and spread.  I have working output at this point.  Basically I pull in and do cdom through all of the particles over all of the runs and then pull out a final frontier, and then I write that frontier out to two txt files in the metrics directory and run the script for spread and the script for hypervolume. 

If you want to give it a shot, just run `sh run.sh` and you'll get the output.  I only have the optimal frontiers in there for Fonseca, and Tanaka at this point.  As they come out of JMetal they're messed up.  Basically they have tabs where they should have spaces.  I have a script called wrangler.py in the spread folder to deal with them, but I can knock those out as we need them.  Take a look and let me know what you think.

I was thinking of adding back in some of the grapher stuff where I could run grapher against the final frontier I'm getting out of the optimizer.  Just haven't quite gotten that far yet.

**Comment by aisobran** [comment]
I just realized there is a bug in the probability implementation. We can end up with negative probs and >1 probs. I'll implement a fix soon, just thought I'd let you guys know. I have a better way to adjust the probabilities by scaling.

**Comment by meneal** [comment]
There was actually a huge evil bug in cdom that took me most of this
morning to fix.  I think it's right now.  I'll need to push before 11:30
whether it's fixed or not.  Hopefully it's ok at that point.

On Thu, Nov 19, 2015 at 10:44 AM, Alexander Sobran <notifications@github.com

**Comment by meneal** [comment]
So I fixed my bug, but for the moment, the only optimizer that will run is adaptive.  I'll fix the others later, this is just because cdom has been changed to return values other than 0/1 in the case that we are some epsilon of loss in difference.  I'm hoping I might be able to talk to Rahul after Data Sci today to get a bit of help to make sure I'm doing some of this right.

**Comment by ghost** [comment]
I pushed my attempts at DTLZ2 and 3, but I noticed a weird bug. If I set m for DTLZ1 to 2 and set m for DTLZ3 to 10, I get a "list index out of range" error on line 218 in Models.py. It turns out that len(x) is correct at 2, but i is 9, which is one less than DTLZ3's m value. So it seems that DTLZ3 is somehow changing the variables of DTLZ1. 

My intuition tells me this has to do with the lambdas that I'm trying to generate for example on line 209. They add just fine with the correct i values, but when they are called later i gets another value. So I'm thinking we need another way of doing this. Do you guys have any suggestions?

**Comment by aisobran** [comment]
Is passing in i from the for loop causing side effects?

**Comment by ghost** [comment]
I think so, but I thought all the instances of i are local and should not cause side effects?

**Comment by aisobran** [comment]
I'll take a deeper dive into it.

All the models were derived from : http://www.tik.ee.ethz.ch/~sop/publicationListFiles/dtlz2001a.pdf
?

**Comment by ghost** [comment]
Yeah. I tried a DTLZ1 m of 5 and it didn't crash, even though DTLZ 3's m is 10. So I'm sure there is some side effect happening but not sure exactly where.

**Comment by ghost** [comment]
Just pushed the fix to DTLZ from class and added DTLZ4. The only addition to DTLZ4 is an alpha exponent. The rest of the problems are more complicated so I didn't want to get started on them right now. I'll definitely get around to them by next weekend when I get back from vacation, otherwise if any of you want to implement them feel free.

If I get time, I'll try to start working on a really rough draft of the intro/related work, but I can really get started on that once I get back next weekend.

I'll have my laptop with me so I can do some small stuff if necessary though.

**Comment by meneal** [comment]
I've added all of the True frontiers now to the True_PF folder in spread.  They should all be in the correct format.  I added all of the DTLZ three dimensional frontiers instead of using the 2d ones.  We could certainly add the 2d as well if necessary.

I'm going to try to do a bit more testing to see how things are working.

**Comment by meneal** [comment]
There's an issue with a few things right now.  
- Running any of the Viennet versions crashes the hypervolume script.  Totally unsure about this at this point.  Also running DTLZ of any sort is crashing this too.  Definitely need to investigate that.
- The patience setup I have at the moment is causing each retry to end at K=401 no idea yet why that's happening.  Maybe there is just nothing changing after that, but I have a hard time believing that we should actually stop after 99 iterations.

**Comment by ghost** [comment]
I just pushed a short beginning to our report. It has Introduction and Related Work sections. I'll flesh these sections out over the course of the weekend.

**Comment by ghost** [comment]
I'm trying to find a reference for the adaptive PSO implementation. The off-the-shelf paper cites "The swarm and the queen: towards a deterministic and adaptive particle swarm optimization" as the one which introduces the constriction factor. Sasha, is this the paper you were referencing when you said "It looks like you are implementing the adaptive PSO by Clerc and Kennedy(2002)..."? The paper I found only has Clerc as an author.

**Comment by aisobran** [comment]
This is the paper I was referencing: http://ieeexplore.ieee.org/xpl/login.jsp?tp=&arnumber=985692&url=http%3A%2F%2Fieeexplore.ieee.org%2Fiel5%2F4235%2F21241%2F00985692

You have to get it through ncsu.lib

**Comment by ghost** [comment]
Thanks Sasha. adaptiveGlobalPSO.py looks similar to the paper you referenced, so I'm thinking we can use your paper as the reference for our adaptive implementation.

Just a note, it looks like the inertia weight factor was introduced in [this paper](http://dsp.szu.edu.cn/pso/ispo/download/a%20modified%20pso.pdf) by Shi and Eberhart. I've added a reference to our report and pushed it. Right now I have divided the related work section into subsections just for clarity, but we can get rid of them later. I'm going to incorporate Sasha's paper into the report, and also a paper about the probability implementations. Is there reference for those you have handy?

**Comment by aisobran** [comment]
The probability implementation is our own heuristic. It basically determines which particle the current particle has had the most success following and sets that as gbest. I can write up the equations used.

**Comment by meneal** [comment]
So a few things on this.  I think we've been really pretty disorganized about how we're going about this and I feel like if we actually want to get it done and get it done in some more sensible way we're going to have to do a bit better for the next little while.  

My thought is we should actually stop posting in this one thread and instead come up with a list of specific features and issues that we want to complete to get this done and have separate github issues for each.  Otherwise, like I said above I think we're going to have problems getting done. 

 So I have a set of things that I think we need and I'll post each into a separate issue and we can discuss each of them in some sort of individual way and we can actually assign someone to take care of it and not be working in a bunch of different directions at once.  

I think a lot of this is my fault, but I also think that we were in a position before this where each homework was so short and really wasn't easy to split up in some sort of sensible way.  So we didn't have to get organized because one person could do a bunch of work and then another person could step in and do a bunch of work and we could just post what we were doing up here.  But I think that the rig is big enough at this point that we're beyond that at this point.
